File: src/validParentheses/Brackets.java
package validParentheses;

import java.util.Stack;

public class Brackets {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char ch : s.toCharArray()) {
            try {
                // Push opening brackets to Stack
                if (ch == '(' || ch == '{' || ch == '[') {
                    stack.push(ch);
                } else if (ch == ')' || ch == '}' || ch == ']') {
                    // Closing brackets in expression must match with Stack's top element
                    if (!isSameType(stack.pop(), ch)) {
                        return false;
                    }
                }
            } catch (Exception e) {
                // Element can't be popped since Stack is empty
                return false;
            }
        }
        return stack.isEmpty();
    }

    private static boolean isSameType(char ch1, char ch2) {
        return ch1 == '(' && ch2 == ')' ||
                ch1 == '{' && ch2 == '}' ||
                ch1 == '[' && ch2 == ']';
    }
}

File: src/validParentheses/brackets.py
class Solution:
    def isValid(self, s: str) -> bool:
        def same(ch1: str, ch2: str) -> bool:
            table = {'(': ')', '{': '}', '[': ']'}
            return table[ch1] == ch2

        stk = []
        for ch in s:
            if ch in "({[":
                stk.append(ch)
            elif ch in ")}]":
                if not (stk and same(stk.pop(), ch)):
                    return False
        return not stk


if __name__ == '__main__':
    s = input("Input: ")
    print(f"Output: {Solution().isValid(s)}")
